plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -82,41 +75,88 @@ public synchronized void refreshAndClearCache(Map<String, GoogleCloudStorageClie
      * method.
      *
      * @param clientName name of the client settings used to create the client
+     * @param repositoryName name of the repository that would use the client
+     * @param stats the stats collector used to gather information about the underlying SKD API calls.
      * @return a cached client storage instance that can be used to manage objects
      *         (blobs)
      */
-    public Storage client(final String clientName) throws IOException {
-        final LazyInitializable<Storage, IOException> lazyClient = clientsCache.get().get(clientName);
-        if (lazyClient == null) {
-            throw new IllegalArgumentException("Unknown client name [" + clientName + "]. Existing client configs: "
-                    + Strings.collectionToDelimitedString(clientsCache.get().keySet(), ","));
+    public Storage client(final String clientName,
+                          final String repositoryName,
+                          final GoogleCloudStorageOperationsStats stats) throws IOException {
+        {
+            final Storage storage = clientCache.get(clientKey(clientName, repositoryName));
+            if (storage != null) {
+                return storage;
+            }
         }
-        return lazyClient.getOrCompute();
+        synchronized (this) {
+            final Storage existing = clientCache.get(clientKey(clientName, repositoryName));
+
+            if (existing != null) {
+                return existing;
+            }
+
+            final GoogleCloudStorageClientSettings settings = clientSettings.get(clientName);
+
+            if (settings == null) {
+                throw new IllegalArgumentException("Unknown client name [" + clientName + "]. Existing client configs: "
+                    + Strings.collectionToDelimitedString(clientSettings.keySet(), ","));
+            }
+
+            logger.debug(() -> new ParameterizedMessage("creating GCS client with client_name [{}], endpoint [{}]", clientName,
+                settings.getHost()));
+            final Storage storage = createClient(settings, stats);
+            clientCache = Maps.copyMapWithAddedEntry(clientCache, clientKey(clientName, repositoryName), storage);
+            return storage;
+        }
+    }
+
+    private String clientKey(String clientName, String repositoryName) {
```

Comment:
I just realized this working on https://github.com/elastic/elasticsearch/pull/56703, do we even need to do this compound key magic? We can just cache the clients by `repositoryName` and it's all good as far as I can see? We never have two repos by the same name, so we can keep this simple right?